src/Interpreters/getColumnFromBlock.cpp

```diff
@@ -31,6 +31,36 @@ ColumnPtr tryGetColumnFromBlock(const Block & block, const NameAndTypePair & req
     return castColumn({elem_column, elem_type, ""}, requested_column.type);
 }
 
+ColumnPtr tryGetSubcolumnFromBlock(const Block & block, const DataTypePtr & requested_column_type, const NameAndTypePair & requested_subcolumn)
```

Comment:
`getColumnFromBlock` is used in a few other places. `tryGetColumnFromBlock` is used in that function. As I couldn't completely understand the changes (I have a lot to learn from Dynamic type), but a question has risen in my mind: aren't the usages of `getColumnFromBlock` can also be problematic?

Reply:
Not all storages support storing dynamic columns. Memory engine supports it and uses `tryGetColumnFromBlock`, so I added special version `tryGetSubcolumnFromBlock`. Other places that use `getColumnFromBlock` should not be a problem.

> As I couldn't completely understand the changes

Let me try to explain. Initially we have column `Dynamic` with default `max_types=32`, and it means that the maximum number of different data types stored in `Dynamic` column is 32. We insert dome UInt64 data in to it, then we do alter modify column and change the type to `Dynamic(max_types=1)`, so now the maximum number of different types is 1, and in this case this single type will be always String, because we can cast any type to it. But in Memory engine when we do such alter we don't change the data, we only change the types in metadata. And when we read something from the data, we just convert column to the new type during reading. So, when we read subcolumn `d.UInt64`, we first read it from old block, get the initial data, and then see that the type of this subcolumn actually didn'g change and don't know about changes in Dynamic type, so we don't cast it to String and get wrong result. I changed the logic, so in case of Dynamic subcolumns we first cast the whole Dynamic column to the new type and only then extract a subcolumn from it.
Hope I could explain it